mesa/experimental/jupyter_viz.py

```diff
@@ -142,44 +143,54 @@ def check_param_is_fixed(param):
         return True
 
 
-def make_user_input(user_input, name, options):
-    """Initialize a user input for configurable model parameters.
+@solara.component
+def UserInputs(user_params, on_change=None):
+    """Initialize user inputs for configurable model parameters.
     Currently supports :class:`solara.SliderInt`, :class:`solara.SliderFloat`,
     and :class:`solara.Select`.
 
-    Args:
-        user_input: :class:`solara.reactive` object with initial value
-        name: field name; used as fallback for label if 'label' is not in options
-        options: dictionary with options for the input, including label,
+    Props:
+        user_params: dictionary with options for the input, including label,
         min and max values, and other fields specific to the input type.
+        on_change: function to be called when the value of an input changes.
     """
-    # label for the input is "label" from options or name
-    label = options.get("label", name)
-    input_type = options.get("type")
-    if input_type == "SliderInt":
-        solara.SliderInt(
-            label,
-            value=user_input,
-            min=options.get("min"),
-            max=options.get("max"),
-            step=options.get("step"),
-        )
-    elif input_type == "SliderFloat":
-        solara.SliderFloat(
-            label,
-            value=user_input,
-            min=options.get("min"),
-            max=options.get("max"),
-            step=options.get("step"),
-        )
-    elif input_type == "Select":
-        solara.Select(
-            label,
-            value=options.get("value"),
-            values=options.get("values"),
-        )
-    else:
-        raise ValueError(f"{input_type} is not a supported input type")
+
+    def handle_change_value(name):
+        def change_value(change):
+            on_change(name, change)
+
+        return change_value
+
+    for name, options in user_params.items():
+        # label for the input is "label" from options or name
+        label = options.get("label", name)
+        input_type = options.get("type")
+        if input_type == "SliderInt":
+            solara.SliderInt(
+                label,
+                value=options.get("value"),
```

Comment:
I meant, `use_reactive`.

Reply:
No strict rule here, but I try to follow a one-directional data flow model here. `JupyterViz` "owns" the state here, i.e. we just pass the current parameters down to `UserInputs`. Any changes are brought back via the `on_change` method and the parent component decides what to do with the changes.

Contrast this with keeping the state inside UserParams. It wouldn't be clear how to extract the current value back to the parent component (JupyterViz) so we can act on a change. I don't know enough about how solara yet, so maybe this would also be possible. I am mainly following React best practices here - for a reference (and a great start to react in general) see https://react.dev/learn/thinking-in-react

Or maybe you meant something completely different?

Reply:
But the question is where should that state live? In the sine example global state is used. Which I did in the previous iteration, where you - correctly - blamed me for it being not reusable that way. 

I think the current way is very simple. At the top level we have the `model_parameters` state variable that is used for the model creation. Its a simple variable-value dictionary. And we have the list of user-settable parameters. This list is passed independently of the `model_parameters` to `UserParams`. And any changes inside `UserParams` are handled at the top level by updating the `model_parameters`. This way we have maximal reusability of the `UserParams` component.

Now if we were to define the reactive variables at the `JupyterViz` component and pass them to `UserParams` we have a tight coupling between the user parameters and the model parameters and the logic resides both in `JupyterViz` and `UserParams`. `UserParams` can't be used without `model_parameters` although that dependency is not strictly required. 

And if we define the reactive variables inside `UserParams` than we are back to my problem of not knowing how to get the updated value back to the parent component. 

I fully agree that it appears a bit convoluted at first and especially the nested functions are not nice - in JS the function definition could simply be written as `(value) => (name) => on_change(name, value)`. I don't know if it would be possible to have a similar concise syntax in Python. But the important part is that it is *not* the same functionality. Just defining `value` as a reactive value doesn`t help us changing the `model_parameters` dictionary. 

 

I could put `use_reactive` inside `UserParams` - but then

Reply:
The nested functions are just one of the complications.
It's that you have to do `handle_*` at the `JupyterViz` level, and do something along the line of subsequent `on_change`'s inside the subcomponents. Simpler would be to have this at the parent component, and at the last leg of the child component.

I'm not sure what you meant by `UserParams`. Is it a hypothetical reactive variable,  a Solara component, a class, or something else not yet implemented? If it is initialized inside `JupyterViz`, why can't `JupyterViz` be notified of its change when you can add do `solara.use_reactive(..., on_change=do_something)`?